runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,757 @@
+	candidate_hash: CandidateHash,
+}
+
+impl DisputesTimeSlot {
+	pub fn new(session_index: SessionIndex, candidate_hash: CandidateHash) -> Self {
+		Self { session_index, candidate_hash }
+	}
+}
+
+/// An offence that is filed when a series of validators lost a dispute.
+#[derive(RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct SlashingOffence<KeyOwnerIdentification> {
+	/// The size of the validator set in that session.
+	pub validator_set_count: ValidatorSetCount,
+	/// Should be unique per dispute.
+	pub time_slot: DisputesTimeSlot,
+	/// Staking information about the validators that lost the dispute
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+	/// What fraction of the total exposure that should be slashed for
+	/// this offence.
+	pub slash_fraction: Perbill,
+	/// Whether the candidate was valid or invalid.
+	pub kind: SlashingOffenceKind,
+}
+
+impl<Offender> Offence<Offender> for SlashingOffence<Offender>
+where
+	Offender: Clone,
+{
+	const ID: Kind = *b"disputes:slashin";
+
+	type TimeSlot = DisputesTimeSlot;
+
+	fn offenders(&self) -> Vec<Offender> {
+		self.offenders.clone()
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.time_slot.session_index
+	}
+
+	fn validator_set_count(&self) -> ValidatorSetCount {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.time_slot.clone()
+	}
+
+	fn disable_strategy(&self) -> DisableStrategy {
+		match self.kind {
+			SlashingOffenceKind::ForInvalid => DisableStrategy::Always,
+			// in the future we might change it based on number of disputes initiated
+			SlashingOffenceKind::AgainstValid => DisableStrategy::Never,
+		}
+	}
+
+	fn slash_fraction(&self, _offenders: u32) -> Perbill {
+		self.slash_fraction
+	}
+}
+
+impl<KeyOwnerIdentification> SlashingOffence<KeyOwnerIdentification> {
+	fn new(
+		session_index: SessionIndex,
+		candidate_hash: CandidateHash,
+		validator_set_count: ValidatorSetCount,
+		offenders: Vec<KeyOwnerIdentification>,
+		kind: SlashingOffenceKind,
+	) -> Self {
+		let time_slot = DisputesTimeSlot::new(session_index, candidate_hash);
+		let slash_fraction = match kind {
+			SlashingOffenceKind::ForInvalid => SLASH_FOR_INVALID,
+			SlashingOffenceKind::AgainstValid => SLASH_AGAINST_VALID,
+		};
+		Self { time_slot, validator_set_count, offenders, slash_fraction, kind }
+	}
+}
+
+/// This type implements `SlashingHandler`.
+pub struct SlashValidatorsForDisputes<C> {
+	_phantom: sp_std::marker::PhantomData<C>,
+}
+
+impl<C> Default for SlashValidatorsForDisputes<C> {
+	fn default() -> Self {
+		Self { _phantom: Default::default() }
+	}
+}
+
+impl<T> SlashValidatorsForDisputes<Pallet<T>>
+where
+	T: Config<KeyOwnerIdentification = IdentificationTuple<T>>,
+{
+	/// If in the current session, returns the identified validators. `None`
+	/// otherwise.
+	fn maybe_identify_validators(
+		session_index: SessionIndex,
+		account_ids: &[AccountId<T>],
```

Comment:
Why not fetch those `AccountId`s directly in this function and only if in the current session?

Reply:
Good idea. Previously, account_ids were used after that function, but now they don't. so this is a remnant from the past.